Events

Strzepowski opened the pull request.

Strzepowski committed:
roles
 /1

Strzepowski committed:
code cosmetics
 /2

Strzepowski committed:
code cosmetics
 /3

Strzepowski committed:
fixed login /w admin logging
    /4

Strzepowski committed:
fixed spring roles recognition
    /5

Strzepowski committed:
fixed spring security authorization
    /6

Strzepowski committed:
quick login fix
    /7

Strzepowski committed:
admin panel work in progress /problems with login validate
    /8

Strzepowski committed:
admin panel work in progress
adjustments to validation
/problems with login navbar
    /9

Strzepowski committed:
admin panel work in progress
problems with login navbar
    /10

Strzepowski committed:
admin panel work in progress
problems with login navbar
    /11

Strzepowski committed:
events work in progress
problems with login navbar disappearing buttons
/1

Strzepowski committed:
events work in progress
problems with login navbar disappearing buttons
/2

Strzepowski committed:
event list on main page
problems with login navbar disappearing buttons and sorting event list
/3

Strzepowski committed:
fixing navbar alignment
event list on main page
problems with login navbar disappearing buttons and sorting event list
/3

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/config/AdminCreationConfiguration.java:
```diff
@@ -0,0 +1,56 @@
+
+            //TODO IS THIS GOOD PLACE TO SET ADMIN LOGIN INFO?
+
+            // SET ADMIN LOGIN
```
STOP YELLING AT PEOPLE :)

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/config/AdminCreationConfiguration.java:
```diff
@@ -0,0 +1,56 @@
+        if (!userService.findByEmail("admin").isPresent()) {
+            RegisterForm adminForm = new RegisterForm();
+
+            //TODO IS THIS GOOD PLACE TO SET ADMIN LOGIN INFO?
```
yup, it's ok

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/config/AdminCreationConfiguration.java:
```diff
@@ -0,0 +1,56 @@
+            // SET ADMIN NICKNAME (SHOWN TO OTHER USERS)
+            adminForm.setNickname("ADMIN");
+            // SET ADMIN PASSWORD
+            adminForm.setPassword("admin");
```
it would be better to generate it randomly (for ex. using `UUID.randomUUID().toString()`) and log that random password so admin can take it from there

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/config/AdminCreationConfiguration.java:
```diff
@@ -0,0 +1,56 @@
+
+@Component
+@Slf4j
+public class AdminCreationConfiguration implements ApplicationListener<ContextRefreshedEvent> {
```
That's the old way of handling events in Spring, change it to `@EventListener` - [take a look here at point 5](https://www.baeldung.com/spring-events)

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/config/SecurityConfiguration.java:
```diff
@@ -31,36 +30,39 @@ protected void configure(HttpSecurity http) throws Exception {
+                    .antMatchers("/", "/register", "/h2db/**", "/login", "/loginForm" )
+                        .permitAll()
+                    .antMatchers("/addEvent")
+                .hasRole("ORGANISER")
```
1. it would be better to format it such way:
`antMatcher(...).hasRole(...)`
2. Another thing is that you may organize your endpoints to somehow reflect roles that are required to access them, for ex.:
all users trying to reach `/organiser/**` are required to have `ORGANISER` role
That's just suggestion, it's your choice

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/controllers/AdminPanelController.java:
```diff
@@ -0,0 +1,23 @@
+    @GetMapping("adminpanel")
+    public ModelAndView get() {
+        ModelAndView modelAndView = new ModelAndView("adminpanel");
+        modelAndView.addObject("allUsers", adminPanelService.findAll());
```
you may think about returning something like DTO instead of whole user (that's just clean code) - layer separation - maybe to be done at polishing stage of development :)

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/controllers/EventController.java:
```diff
@@ -0,0 +1,48 @@
+
+    @PostMapping("/addEvent")
+    public ModelAndView addEventPost(@ModelAttribute EventForm eventForm, BindingResult bindingResult) {
+        eventValidator.validate(eventForm, bindingResult);
```
do that properly, instead of manually calling `validate` use `@Valid` annotation - you can hook up your validator for ex. like [that](http://dolszewski.com/spring/custom-validation-annotation-in-spring/) or [that](https://www.concretepage.com/spring/spring-mvc/spring-mvc-custom-validator) - it applies to all controllers, don't call `validate` manually, use Spring's mechanism.

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/controllers/EventController.java:
```diff
@@ -0,0 +1,48 @@
+import pl.sda.eventmanager.services.EventValidator;
+
+@Controller
+public class EventController extends IndexListController {
```
I don't see any benefit coming from extending `IndexListController`, please revise usability of `IndexListController` and `IndexController`

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/controllers/LoginController.java:
```diff
@@ -7,36 +7,43 @@
 
-    @PostMapping("login")
-    public ModelAndView loginPost(@ModelAttribute LoginForm loginForm, BindingResult bindingResult) {
+        if(loginForm.getEmail().equals("admin")){
```
I would rather avoid writing that `if`, maybe just set admin mail to some temporary not valid mail and after first login force him to change it? or even better create login site exclusively for first login (without such restrictive mail validation) and then force him to change it. Quite few ways to solve that problem, however leave it for later (first core functionality).

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/dto/RegisterForm.java:
```diff
@@ -10,5 +10,5 @@
     private String password;
     private String confirmPassword;
-    private Role role = Role.USER;
+    private Role role = Role.ROLE_USER;
```
in short create interface that `UserRegisterForm`(this class) and `AdminRegisterParameters`(class that you can use to register admin) will implement, interface should consist only of methods necessary to register new User/Admin (getters), in `UserRegisterForm` you can hardcode `getRole()` to `Role.ROLE_USER` and in `AdminRegisterParameters` to `Role.ROLE_ADMI`N - more on saturday (f2f) - for optional implementation later

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/model/Event.java:
```diff
@@ -0,0 +1,105 @@
+
+
+
+    public static final class EventBuilder {
```
no need to create that manually, you can use `@Builder` on `Event`, if you decide to keep that class for whatever reason, please remove unused methods `anEvent()`

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/model/Event.java:
```diff
@@ -0,0 +1,105 @@
+
+
+    public static final class EventBuilder {
+        private Long Id;
```
`Id` should be lowercase

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/model/User.java:
```diff
@@ -26,22 +23,47 @@
+    @JoinTable(name = "USER_EVENT",
+            joinColumns = {@JoinColumn(name = "EVENT_ID", nullable = false)},
+            inverseJoinColumns = {@JoinColumn(name = "USER_ID", nullable = false)})
+    private Set<Event> attendedEvents = new HashSet<>();
```
no need to initialize it

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/model/User.java:
```diff
@@ -26,22 +23,47 @@
 
-    @OneToOne(mappedBy = "user", cascade = CascadeType.ALL, orphanRemoval = true)
+    @OneToMany(mappedBy = "eventOrganiser", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
+    private Set<Event> organisedEvents = new HashSet<>();
```
no need to initialize it

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/model/User.java:
```diff
@@ -26,22 +23,47 @@
     public Collection<? extends GrantedAuthority> getAuthorities() {
 
-        return Collections.singletonList(new SimpleGrantedAuthority(Role.USER.name()));
+        List<SimpleGrantedAuthority> organiserAuthorityList = new ArrayList<>();
```
move this list to const

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/model/User.java:
```diff
@@ -26,22 +23,47 @@
+        organiserAuthorityList.add(new SimpleGrantedAuthority(Role.ROLE_USER.name()));
+        organiserAuthorityList.add(new SimpleGrantedAuthority(Role.ROLE_ORGANISER.name()));
+
+        List<SimpleGrantedAuthority> adminAuthorityList = new ArrayList<>();
```
move this list to const

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/model/User.java:
```diff
@@ -67,13 +89,14 @@ public boolean isEnabled() {
     }
 
+
     public static final class UserBuilder {
```
you can use lombok's `@Builder` instead of writing that manually

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/model/Event.java:
```diff
@@ -0,0 +1,105 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long Id;
```
`Id` should be lowercase

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/services/AdminPanelService.java:
```diff
@@ -0,0 +1,24 @@
+
+    public List<User> findAll(){
+        List<User> allByOrderById = userRepository.findAllByOrderById();
+        allByOrderById.remove(0);
```
1. I don't think you need to remove admin from user list.
2. To remove admin from list find him exclude him by role (using proper Spring Data method). Right now for someone who doesn't know logic of your app it's hard to understand why you're removing user with `id = 0`

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/services/AdminPanelService.java:
```diff
@@ -0,0 +1,24 @@
+        this.userRepository = userRepository;
+    }
+
+    public List<User> findAll(){
```
rather `findAllWithoutAdmin()`

red-rocket-dev replied:
move that method to `UserService` (or remove if there is already method `findAll()`), as pointed below you don't need to exclude admin

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/services/AdminPanelService.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.List;
+
+@Service
+public class AdminPanelService {
```
Better name would be `AdminOperationsService`/`AdminActionsService`, provided that you will keep there methods that are **ONLY** performed by admin and are not performed anywhere else

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/services/AdminValidator.java:
```diff
@@ -0,0 +1,37 @@
+
+
+        ValidationUtils.rejectIfEmptyOrWhitespace(errors, "password", "NotEmpty", "Please use proper email address.");
+        if (!passwordEncoder.matches(loginForm.getPassword(), userService.findByEmail(loginForm.getEmail()).get().getPassword())) {
```
leave that to Spring Security, don't check mail and/or password manually

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/services/EventService.java:
```diff
@@ -0,0 +1,60 @@
+        return eventRepository.findByEventName(eventName);
+    }
+
+    public List<Event> findAllOngoingAndFutureEvents() {
```
so..just `findAllEvents`?

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/services/EventService.java:
```diff
@@ -0,0 +1,60 @@
+
+    //TODO LIST IS NOT SORTED ON "/"
+
+    public List<Event> sortEventList(List<Event> events) {
```
all sorting related to objects stored in DB should be done using DB (for ex. replace with findBy..OrderByEventEndDesc)

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/services/LoginValidator.java:
```diff
@@ -0,0 +1,43 @@
+            errors.rejectValue("email", "WrongEmailOrPassword", "Wrong email or password.");
+        } else {
+            ValidationUtils.rejectIfEmptyOrWhitespace(errors, "password", "NotEmpty", "All fields are required.");
+            if (!passwordEncoder.matches(loginForm.getPassword(), userService.findByEmail(loginForm.getEmail()).get().getPassword())) {
```
same as with `AdminValidatorLogin`, don't check that, leave that to Spring Security

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/services/RegisterValidator.java:
```diff
@@ -0,0 +1,53 @@
+        RegisterForm registerForm = (RegisterForm) o;
+
+        ValidationUtils.rejectIfEmptyOrWhitespace(errors, "email", "NotEmpty", "All fields are required.");
+        if (userService.findByEmail(registerForm.getEmail()).isPresent()) {
```
currently you're performing 2 SQL queries to check email and nickname uniqueness, this can be reduced to 1, more f2f.

red-rocket-dev commented on src/main/java/pl/sda/eventmanager/services/UserService.java:
```diff
@@ -25,30 +28,33 @@ public UserService(UserRepository userRepository, PasswordEncoder passwordEncode
     @Override
     @Transactional(readOnly = true)
     public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        if(userRepository.findByEmail(email).isPresent())
```
use brackets (`{}`) at **all** times, that improves code readability and decreases chance of error while modifying `if`

red-rocket-dev replied:
using `isPresent()` is bad practice (with some exceptions), refactor code to use one of it's methods.

red-rocket-dev commented on src/main/resources/validation.properties:
```diff
@@ -1,11 +1,15 @@
-Size.registerForm.nickname=Please use between 3 and 50 characters.
 Duplicate.registerForm.nickname=Username taken. Please choose another username.
 Duplicate.registerForm.email=Email already in base. Please Create account on another email.
+Duplicate.eventForm.eventName=Event name already in our database. Please choose a different event name.
```
nice

red-rocket-dev reviewed (changes requested):
Most of comments are not critical, so can be resolved later, but some of them should be resolved ASAP :)

Strzepowski committed:
fixing navbar alignment
event list on main page
problems with login navbar disappearing buttons and sorting event list
/4

Strzepowski committed:
search event query

problems with login navbar disappearing buttons and sorting event list
/5

Strzepowski committed:
search bar ok,

problems with login navbar disappearing buttons and sorting event list
/6

Strzepowski committed:
event details page done, minor bugfix

problems with login navbar disappearing buttons and sorting event list
/7

Strzepowski committed:
admin can't delete admin account

Strzepowski committed:
searched event name persists to next page

Strzepowski committed:
working comments functionality

Strzepowski committed:
bugfix
